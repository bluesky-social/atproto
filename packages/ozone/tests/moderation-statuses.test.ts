import assert from 'node:assert'
import {
  TestNetwork,
  SeedClient,
  basicSeed,
  ModeratorClient,
} from '@atproto/dev-env'
import {
  ToolsOzoneModerationDefs,
  ToolsOzoneModerationQueryStatuses,
} from '@atproto/api'
import { forSnapshot } from './_util'
import {
  REASONMISLEADING,
  REASONSPAM,
} from '../src/lexicon/types/com/atproto/moderation/defs'
import {
  REVIEWOPEN,
  REVIEWNONE,
} from '../src/lexicon/types/tools/ozone/moderation/defs'

describe('moderation-statuses', () => {
  let network: TestNetwork
  let sc: SeedClient
  let modClient: ModeratorClient

  const seedEvents = async () => {
    const bobsAccount = {
      $type: 'com.atproto.admin.defs#repoRef',
      did: sc.dids.bob,
    }
    const carlasAccount = {
      $type: 'com.atproto.admin.defs#repoRef',
      did: sc.dids.alice,
    }
    const bobsPost = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.bob][0].ref.uriStr,
      cid: sc.posts[sc.dids.bob][0].ref.cidStr,
    }
    const alicesPost = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.alice][1].ref.uriStr,
      cid: sc.posts[sc.dids.alice][1].ref.cidStr,
    }

    for (let i = 0; i < 4; i++) {
      await sc.createReport({
        reasonType: i % 2 ? REASONSPAM : REASONMISLEADING,
        reason: 'X',
        //   Report bob's account by alice and vice versa
        subject: i % 2 ? bobsAccount : carlasAccount,
        reportedBy: i % 2 ? sc.dids.alice : sc.dids.bob,
      })
      await sc.createReport({
        reasonType: REASONSPAM,
        reason: 'X',
        //   Report bob's post by alice and vice versa
        subject: i % 2 ? bobsPost : alicesPost,
        reportedBy: i % 2 ? sc.dids.alice : sc.dids.bob,
      })
    }
  }

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'ozone_moderation_statuses',
    })
    sc = network.getSeedClient()
    modClient = network.ozone.getModClient()
    await basicSeed(sc)
    await network.processAll()
    await seedEvents()
  })

  afterAll(async () => {
    await network.close()
  })

  describe('query statuses', () => {
    it('returns statuses for subjects that received moderation events', async () => {
      const response = await modClient.queryStatuses({})

      expect(forSnapshot(response.subjectStatuses)).toMatchSnapshot()
    })

    it('returns statuses filtered by subject language', async () => {
      const klingonQueue = await modClient.queryStatuses({
        tags: ['lang:i'],
      })

      expect(forSnapshot(klingonQueue.subjectStatuses)).toMatchSnapshot()

      const nonKlingonQueue = await modClient.queryStatuses({
        excludeTags: ['lang:i'],
      })

      // Verify that the klingon tagged subject is not returned when excluding klingon
      expect(nonKlingonQueue.subjectStatuses.map((s) => s.id)).not.toContain(
        klingonQueue.subjectStatuses[0].id,
      )

      // Verify multi lang tag exclusion
      Promise.all(
        nonKlingonQueue.subjectStatuses.map((s, i) => {
          return modClient.emitEvent({
            subject: s.subject,
            event: {
              $type: 'tools.ozone.moderation.defs#modEventTag',
              add: [i % 2 ? 'lang:jp' : 'lang:it'],
              remove: [],
              comment: 'Adding custom lang tag',
            },
            createdBy: sc.dids.alice,
          })
        }),
      )

      const queueWithoutKlingonAndItalian = await modClient.queryStatuses({
        excludeTags: ['lang:i', 'lang:it'],
      })

      queueWithoutKlingonAndItalian.subjectStatuses
        .map((s) => s.tags)
        .flat()
        .forEach((tag) => {
          expect(['lang:it', 'lang:i']).not.toContain(tag)
        })
    })

    it('returns paginated statuses', async () => {
      // We know there will be exactly 4 statuses in db
      const getPaginatedStatuses = async (
        params: ToolsOzoneModerationQueryStatuses.QueryParams,
      ) => {
        let cursor: string | undefined = ''
        const statuses: ToolsOzoneModerationDefs.SubjectStatusView[] = []
        let count = 0
        do {
          const results = await modClient.queryStatuses({
            limit: 1,
            cursor,
            ...params,
          })
          cursor = results.cursor
          statuses.push(...results.subjectStatuses)
          count++
          // The count is just a brake-check to prevent infinite loop
        } while (cursor && count < 10)

        return statuses
      }

      const list = await getPaginatedStatuses({})
      expect(list[0].id).toEqual(7)
      expect(list[list.length - 1].id).toEqual(1)

      await modClient.emitEvent({
        subject: list[1].subject,
        event: {
          $type: 'tools.ozone.moderation.defs#modEventAcknowledge',
          comment: 'X',
        },
      })

      const listReviewedFirst = await getPaginatedStatuses({
        sortDirection: 'desc',
        sortField: 'lastReviewedAt',
      })

      // Verify that the item that was recently reviewed comes up first when sorted descendingly
      // while the result set always contains same number of items regardless of sorting
      expect(listReviewedFirst[0].id).toEqual(list[1].id)
      expect(listReviewedFirst.length).toEqual(list.length)
    })
  })

  describe('reviewState changes', () => {
    it('only sets state to #reviewNone on first non-impactful event', async () => {
      const bobsAccount = {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.bob,
      }
      const alicesPost = {
        $type: 'com.atproto.repo.strongRef',
        uri: sc.posts[sc.dids.alice][0].ref.uriStr,
        cid: sc.posts[sc.dids.alice][0].ref.cidStr,
      }
      const getBobsAccountStatus = async () => {
        const data = await modClient.queryStatuses({
          subject: bobsAccount.did,
        })

        return data.subjectStatuses[0]
      }
      // Since bob's account already had a reviewState, it won't be changed by non-impactful events
      const bobsAccountStatusBeforeTag = await getBobsAccountStatus()

      await Promise.all([
        modClient.emitEvent({
          subject: bobsAccount,
          event: {
            $type: 'tools.ozone.moderation.defs#modEventTag',
            add: ['newTag'],
            remove: [],
            comment: 'X',
          },
          createdBy: sc.dids.alice,
        }),
        modClient.emitEvent({
          subject: bobsAccount,
          event: {
            $type: 'tools.ozone.moderation.defs#modEventComment',
            comment: 'X',
          },
          createdBy: sc.dids.alice,
        }),
      ])
      const bobsAccountStatusAfterTag = await getBobsAccountStatus()

      expect(bobsAccountStatusBeforeTag.reviewState).toEqual(
        bobsAccountStatusAfterTag.reviewState,
      )

      // Since alice's post didn't have a reviewState it is set to reviewNone on first non-impactful event
      const getAlicesPostStatus = async () => {
        const data = await modClient.queryStatuses({
          subject: alicesPost.uri,
        })

        return data.subjectStatuses[0]
      }

      const alicesPostStatusBeforeTag = await getAlicesPostStatus()
      expect(alicesPostStatusBeforeTag).toBeUndefined()

      await modClient.emitEvent({
        subject: alicesPost,
        event: {
          $type: 'tools.ozone.moderation.defs#modEventComment',
          comment: 'X',
        },
        createdBy: sc.dids.alice,
      })
      const alicesPostStatusAfterTag = await getAlicesPostStatus()
      expect(alicesPostStatusAfterTag.reviewState).toEqual(REVIEWNONE)

      await modClient.emitEvent({
        subject: alicesPost,
        event: {
          $type: 'tools.ozone.moderation.defs#modEventReport',
          reportType: REASONMISLEADING,
          comment: 'X',
        },
        createdBy: sc.dids.alice,
      })
      const alicesPostStatusAfterReport = await getAlicesPostStatus()
      expect(alicesPostStatusAfterReport.reviewState).toEqual(REVIEWOPEN)
    })
  })

  describe('blobs', () => {
    it('are tracked on takendown subject', async () => {
      const post = sc.posts[sc.dids.carol][0]
      assert(post.images.length > 1)
      await modClient.emitEvent({
        event: {
          $type: 'tools.ozone.moderation.defs#modEventTakedown',
        },
        subject: {
          $type: 'com.atproto.repo.strongRef',
          uri: post.ref.uriStr,
          cid: post.ref.cidStr,
        },
        subjectBlobCids: [post.images[0].image.ref.toString()],
        createdBy: sc.dids.alice,
      })
      const result = await modClient.queryStatuses({
        subject: post.ref.uriStr,
      })
      expect(result.subjectStatuses.length).toBe(1)
      expect(result.subjectStatuses[0]).toMatchObject({
        takendown: true,
        subjectBlobCids: [post.images[0].image.ref.toString()],
      })
    })

    it('are tracked on reverse-takendown subject based on previous status', async () => {
      const post = sc.posts[sc.dids.carol][0]
      await modClient.emitEvent({
        event: {
          $type: 'tools.ozone.moderation.defs#modEventReverseTakedown',
        },
        subject: {
          $type: 'com.atproto.repo.strongRef',
          uri: post.ref.uriStr,
          cid: post.ref.cidStr,
        },
      })
      const result = await modClient.queryStatuses({
        subject: post.ref.uriStr,
      })
      expect(result.subjectStatuses.length).toBe(1)
      expect(result.subjectStatuses[0]).toMatchObject({
        takendown: false,
        subjectBlobCids: [post.images[0].image.ref.toString()],
      })
    })
  })
})
