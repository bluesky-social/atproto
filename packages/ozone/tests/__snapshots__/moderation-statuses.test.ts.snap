// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moderation-statuses query statuses returns statuses filtered by subject language 1`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "hosting": Object {
      "$type": "tools.ozone.moderation.defs#recordHosting",
      "status": "unknown",
    },
    "id": 7,
    "lastReportedAt": "1970-01-01T00:00:00.000Z",
    "reviewState": "tools.ozone.moderation.defs#reviewOpen",
    "subject": Object {
      "$type": "com.atproto.repo.strongRef",
      "cid": "cids(0)",
      "uri": "record(0)",
    },
    "subjectBlobCids": Array [],
    "subjectRepoHandle": "bob.test",
    "tags": Array [
      "lang:en",
      "lang:i",
    ],
    "takendown": false,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "hosting": Object {
      "$type": "tools.ozone.moderation.defs#accountHosting",
      "status": "unknown",
    },
    "id": 5,
    "lastReportedAt": "1970-01-01T00:00:00.000Z",
    "reviewState": "tools.ozone.moderation.defs#reviewOpen",
    "subject": Object {
      "$type": "com.atproto.admin.defs#repoRef",
      "did": "user(0)",
    },
    "subjectBlobCids": Array [],
    "subjectRepoHandle": "bob.test",
    "tags": Array [
      "lang:en",
      "lang:i",
    ],
    "takendown": false,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
]
`;

exports[`moderation-statuses query statuses returns statuses for subjects that received moderation events 1`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "hosting": Object {
      "$type": "tools.ozone.moderation.defs#recordHosting",
      "status": "unknown",
    },
    "id": 7,
    "lastReportedAt": "1970-01-01T00:00:00.000Z",
    "reviewState": "tools.ozone.moderation.defs#reviewOpen",
    "subject": Object {
      "$type": "com.atproto.repo.strongRef",
      "cid": "cids(0)",
      "uri": "record(0)",
    },
    "subjectBlobCids": Array [],
    "subjectRepoHandle": "bob.test",
    "tags": Array [
      "lang:en",
      "lang:i",
    ],
    "takendown": false,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "hosting": Object {
      "$type": "tools.ozone.moderation.defs#accountHosting",
      "status": "unknown",
    },
    "id": 5,
    "lastReportedAt": "1970-01-01T00:00:00.000Z",
    "reviewState": "tools.ozone.moderation.defs#reviewOpen",
    "subject": Object {
      "$type": "com.atproto.admin.defs#repoRef",
      "did": "user(0)",
    },
    "subjectBlobCids": Array [],
    "subjectRepoHandle": "bob.test",
    "tags": Array [
      "lang:en",
      "lang:i",
    ],
    "takendown": false,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "hosting": Object {
      "$type": "tools.ozone.moderation.defs#recordHosting",
      "status": "unknown",
    },
    "id": 3,
    "lastReportedAt": "1970-01-01T00:00:00.000Z",
    "reviewState": "tools.ozone.moderation.defs#reviewOpen",
    "subject": Object {
      "$type": "com.atproto.repo.strongRef",
      "cid": "cids(1)",
      "uri": "record(1)",
    },
    "subjectBlobCids": Array [],
    "subjectRepoHandle": "alice.test",
    "tags": Array [
      "lang:ha",
    ],
    "takendown": false,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "hosting": Object {
      "$type": "tools.ozone.moderation.defs#accountHosting",
      "status": "unknown",
    },
    "id": 1,
    "lastReportedAt": "1970-01-01T00:00:00.000Z",
    "reviewState": "tools.ozone.moderation.defs#reviewOpen",
    "subject": Object {
      "$type": "com.atproto.admin.defs#repoRef",
      "did": "user(1)",
    },
    "subjectBlobCids": Array [],
    "subjectRepoHandle": "alice.test",
    "tags": Array [
      "lang:und",
    ],
    "takendown": false,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
]
`;
